### PR TITLE
Adds a is_null() check to CStr::from_ptr()

### PR DESCRIPTION
**Pull request by mprovost**
Return an empty string instead of segfaulting on strlen().

### CODE REVIEW

**Comment by sfackler** [comment]
Silently converting one invalid pointer to an empty string doesn't seem like the right behavior. There is not an empty string at the address 0 any more than there's an empty string (probably) at the address 6.

**Comment by TimNN** [comment]
I'm gonna re-assign this to the libs team, since they already replied, let's

r? @sfackler 

---

However my initial reaction would have been the same: silently converting to an empty string doesn't seem like the right behavior, although given the return type, there is probably little else that can be done.

Given that the function is `unsafe`, panicking also doesn't feel right, however maybe a `debug_assertion!` would be warranted?

---

Reading your motivation, I'm wondering if you have encountered the behavior (`libc` returning `null`, so that `CStr::from_ptr` segfaults) in the standard library (that would definitely be another issue worth fixing)?

**Comment by mprovost** [comment]
Thanks for having a look. I came across this when writing tests for getlogin(). I was calling it directly from libc but looking at the src of the libc crate it's just importing getlogin() as an extern and isn't checking for nulls either. This could be fixed in the crate for the functions that can return nulls but that just pushes the problem up a level, they're still limited to returning a *c_char which presumably will be attempted to be turned into a CStr at some point. It would be better for these functions to return Options but that's obviously a bigger change. It's easy enough to write a safe wrapper that checks for null and returns None but I was thinking this might be better handled at this layer rather than having lots of if_null() calls, if folk even remember to check. I can put in a debug_assertion! if people think that's the way to go instead.

**Comment by TimNN** [comment]
>  I came across this when writing tests for `getlogin()`. I was calling it directly from libc but looking at the src of the libc crate it's just importing `getlogin()` as an `extern` and isn't checking for nulls either.

> I was thinking this might be better handled at this layer rather than having lots of `if_null()` calls, if folk even remember to check

Note that this is Working As Intended in my opinion: `libc` does exactly what the name implies, it provides access to the types and functions found in `libc`. It is the programmers responsibility to understand the functions and their return values.

If one is using `libc`, one needs to think a bit like a C-programmer, which also means explicitly checking for `null` where appropriate. 

Providing a more Rust-idiomatic API would probably be the responsibility of some higher-level crate, similar to what [`nix`](https://docs.rs/nix/) does for UNIX APIs.

**Comment by sfackler** [comment]
A debug_assert won't do anything since the standard library is built with those disabled.

Again, even a version of `new` that returned an Option would still be fundamentally unsafe. Almost all addresses are not valid strings!

**Comment by mprovost** [comment]
Thanks for the feedback. It sounds like the best thing to do is to write a crate that wraps these functions safely into an Option and checks for null. It's such a mechanical check that it could probably be made into a macro.

Of course from_ptr() can't ever be made totally safe for all input pointers, but well behaved functions either return a pointer to a valid string or null and it currently only handles one of those possibilities. And it feels like segfaulting on the null path isn't very ergonomic. It would be better to at least catch the null and panic with a more friendly error message - I can change this patch to do that instead if anyone else thinks it's a good idea. There's not much point in calling strlen() just to get a core.

It's a shame that there's no better way to handle both common inputs though.

File: src/libstd/ffi/c_str.rs [hunk 1]
```diff
@@ -970,9 +970,13 @@ impl CStr {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     pub unsafe fn from_ptr<'a>(ptr: *const c_char) -> &'a CStr {
-        let len = sys::strlen(ptr);
-        let ptr = ptr as *const u8;
-        CStr::from_bytes_with_nul_unchecked(slice::from_raw_parts(ptr, len as usize + 1))
+        if ptr.is_null() {
```
Comment (TimNN) [inline src/libstd/ffi/c_str.rs]: nit: this can be written as `assert!(!ptr.is_null(), "msg...")`.

**Comment by TimNN** [comment]
Ping form triage @sfackler / @rust-lang/libs: The PR changes a probably guaranteed segfault to a panic, what do you think? It technically is a breaking change.

**Comment by alexcrichton** [comment]
It sounds like the conclusion here was to create a wrapper crate? Should this be closed?

**Comment by mprovost** [comment]
The idea for the crate is to wrap the libc functions in something that returns an Option instead of a raw pointer but that's a more substantial piece of work. I just thought that while I was in the libc module that adding a check for a null pointer would make it more user friendly with a clear panic message instead of segfaulting in strlen().

**Comment by alexcrichton** [comment]
Sure yeah, and @sfackler has provided counterarguments, and it [sounds like you agree](https://github.com/rust-lang/rust/pull/55790#issuecomment-437288247). Do you disagree and would like to still merge this PR?

**Comment by mprovost** [comment]
I agree that returning an empty string is the wrong behaviour, and that changing the return type to an Option belongs in a crate. This is a simpler check that panics instead of segfaulting, I still think that's a valuable usability improvement. I guess the downside is that it's doing an extra check for null every time you create a CStr from a pointer? And if you're already doing that check or know that it can't be null then it's wasted effort. I can't imagine that anyone is relying on the current behaviour where it segfaults. I suppose another option would be to document the current behaviour? The docs for from_ptr() already have a list of reasons why it's unsafe, I could add "ptr is not checked for null" or something. Thanks for looking anyway!

**Comment by sfackler** [comment]
Null inputs seem to be encompassed under the "There is no guarantee to the validity of ptr." bit of the docs.

**Comment by mprovost** [comment]
Ok cool, let's close this then. Thanks for taking the time to consider it!

**Pull request closed without merging**